earth_enterprise/rpms/test/key-value_test.sh

```diff
@@ -0,0 +1,203 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Unit tests for the key value store used by the RPMs.  The functions that
+# start with "test_" are the unit tests.  Other functions are utility
+# functions.
+
+# Load the functions to test
+source ../shared/lib/key-value.sh
+
+test_keyvalue_file_string_starts_with() {
+  assertTrue "Positive key value starts with failed" "keyvalue_file_string_starts_with abcde abc"
+  assertFalse "Negative key value starts with failed" "keyvalue_file_string_starts_with abcde xyz"
+  assertTrue "Empty string starts with failed" "keyvalue_file_string_starts_with abcde ''"
+  assertTrue "Same string starts with failed" "keyvalue_file_string_starts_with abcde abcde"
+  assertFalse "Prefix too long starts with failed" "keyvalue_file_string_starts_with abcde abcdef"
+}
+
+test_keyvalue_file_string_ends_with() {
+  assertTrue "Positive key value ends with failed" "keyvalue_file_string_ends_with abcde cde"
+  assertFalse "Negative key value ends with failed" "keyvalue_file_string_ends_with abcde xyz"
+  assertTrue "Empty string ends with failed" "keyvalue_file_string_ends_with abcde ''"
+  assertTrue "Same string ends with failed" "keyvalue_file_string_ends_with abcde abcde"
+  assertFalse "Prefix too long ends with failed" "keyvalue_file_string_ends_with abcde abcdef"
+}
+
+function_with_output_test() {
+  # $1 - function name
+  # $2 - message
+  # $3 - input value
+  # $4 - expected output
+  "$1" "$3" "OUTPUT"
+  assertEquals "$2" "$4" "$OUTPUT"
+}
+
+key_escape_test() {
+  # $1 - message
+  # $2 - input value
+  # $3 - expected output
+  function_with_output_test "keyvalue_file_escape_key" "$1" "$2" "$3"
+}
+
+test_keyvalue_file_escape_key() {
+  key_escape_test "Basic key escape failed" "abcde" "abcde"
+  key_escape_test "Space key escape failed" "hello world" "hello\ world"
+  key_escape_test "Dollar key escape failed" 'Owed $20' 'Owed\ \$20'
+  key_escape_test "Backslash key escape failed" 'back\slash' 'back\\slash'
+  key_escape_test "Multiple key escape failed" ' \$' '\ \\\$'
+}
+
+value_escape_test() {
+  # $1 - message
+  # $2 - input value
+  # $3 - expected output
+  function_with_output_test "keyvalue_file_escape_value" "$1" "$2" "$3"
+}
+
+test_keyvalue_file_escape_value() {
+  value_escape_test "Basic value escape failed" "abcde" "abcde"
+  value_escape_test "New line value escape failed" 'abc
+def' 'abc\ndef'
+  value_escape_test "Carriage return value escape failed" $'abc\rdef' 'abc\rdef'
+  value_escape_test "Backslash value escape failed" 'back\slash' 'back\\slash'
+  value_escape_test "Backspace value escape failed" $'abc\bdef' 'abc\bdef'
+  value_escape_test "Form feed value escape failed" $'abc\fdef' 'abc\fdef'
+  value_escape_test "Vertical tab value escape failed" $'abc\vdef' 'abc\vdef'
+  value_escape_test "Multiple value escape failed" $'\r\n\\\b\f\v' '\r\n\\\b\f\v'
```

Comment:
Note: You may want to test the case of a value ending with one, or more new-line characters.